crates/alloy/src/reth_db_layer.rs

```diff
@@ -0,0 +1,72 @@
+use alloy_eips::{BlockId, BlockNumberOrTag};
+use eyre::Result;
+use reth_db::DatabaseEnv;
+use reth_node_ethereum::EthereumNode;
+use reth_node_types::NodeTypesWithDBAdapter;
+use reth_provider::{
+    BlockNumReader, DatabaseProviderFactory, ProviderError, ProviderFactory, StateProvider,
+    TryIntoHistoricalStateProvider,
+};
+use std::{path::PathBuf, sync::Arc};
+
+/// A tower-like layer that should be used as a [ProviderLayer](https://docs.rs/alloy/latest/alloy/providers/trait.ProviderLayer.html) in alloy when wrapping the
+/// [Provider](https://docs.rs/alloy/latest/alloy/providers/trait.Provider.html) trait over reth-db.
+#[derive(Debug, Clone)]
+pub struct RethDbLayer {
+    db_path: PathBuf,
+}
+
+impl RethDbLayer {
+    /// Initialize the `RethDbLayer` with the path to the reth datadir.
+    pub const fn new(db_path: PathBuf) -> Self {
+        Self { db_path }
+    }
+
+    /// Get the provided path.
+    pub const fn db_path(&self) -> &PathBuf {
+        &self.db_path
+    }
+}
+
+/// A helper type to get the appropriate DB provider.
+#[derive(Debug, Clone)]
+pub struct DbAccessor<DB = ProviderFactory<NodeTypesWithDBAdapter<EthereumNode, Arc<DatabaseEnv>>>>
+where
+    DB: DatabaseProviderFactory<Provider: TryIntoHistoricalStateProvider + BlockNumReader>,
+{
```

Comment:
do we even need these concrete types here @klkvr ?

I assume this could actually operate entirely on stacked provider traits